Fix ambiguous `Const` instances and add missing instances

joroKr21 opened the pull request:
- Reorder instances in `Const` systematically
 - Fix ambiguous `InvariantMonoidal` instance
 - Add `ContravariantSemigroupal` instance
 - Deprecate unnecessary `Functor` instance
 - Add `Hash` instance and `hash` method
 - Add `SemigroupK` and `MonoidK` instances

satorg commented on core/src/main/scala/cats/data/Const.scala:
```diff
@@ -39,6 +41,7 @@ final case class Const[A, B](getConst: A) {
   def combine(that: Const[A, B])(implicit A: Semigroup[A]): Const[A, B] =
     Const(A.combine(getConst, that.getConst))
 
+  @nowarn("cat=unused")
   def traverse[F[_], C](f: B => F[C])(implicit F: Applicative[F]): F[Const[A, C]] =
     F.pure(retag[C])
```
Just a hint (feel free to disregard): with help of [scalac-compat](https://github.com/typelevel/scalac-compat) it can be addressed in a nicer way:
```
import org.typelevel.scalaccompat.annotation._
```
then
```suggestion
  def traverse[F[_], C](@unused f: B => F[C])(implicit F: Applicative[F]): F[Const[A, C]] =
    F.pure(retag[C])
```

joroKr21 replied:
```scala
[error] /Users/georgy/work/typelevel/cats/core/src/main/scala/cats/data/Const.scala:26:12: object typelevel is not a member of package org
[error] import org.typelevel.scalaccompat.annotation._
[error]            ^
[error] /Users/georgy/work/typelevel/cats/core/src/main/scala/cats/data/Const.scala:46:26: not found: type unused
[error]   def traverse[F[_], C](@unused f: B => F[C])(implicit F: Applicative[F]): F[Const[A, C]] =
[error]         
```

satorg replied:
Yeah, I guess it should work better once these changes (typelevel/sbt-typelevel#518) will be available in the current version of the plugin 🤷 

Meanwhile you could add it to `build.sbt` temporarily:
```sbt
ThisBuild / libraryDependencies += "org.typelevel" %% "scalac-compat-annotation" % "0.1.0" % CompileTime
```
... or skip my initial comment about it entirely ...

joroKr21 replied:
We have a lot of unused (and other warnings in the repo). I guess it would be good to tackle them together.

satorg commented on core/src/main/scala/cats/data/Const.scala:
```diff
@@ -174,44 +200,48 @@ sealed abstract private[data] class ConstInstances0 extends ConstInstances1 {
+sealed abstract private[data] class ConstInstances3 extends ConstInstances4 {
 
-  implicit def catsDataFunctorForConst[C]: Functor[Const[C, *]] =
+  def catsDataFunctorForConst[C]: Functor[Const[C, *]] =
```
Perhaps, it makes sense to mark it `@deprecated` as well.

joroKr21 replied:
Sure, but since which version?

armanbilge replied:
The next release will be Cats 2.10.0. So since then.

satorg commented:
@joroKr21 :
> Fix ambiguous InvariantMonoidal / instance
> Add InvariantSemigroupal instance

Cannot find those in your changes 🤔 
Did you mean `ContravariantMonoidal` / `ContravariantSemigroupal` ?

joroKr21 commented:
> Cannot find those in your changes 🤔
Did you mean ContravariantMonoidal / ContravariantSemigroupal ?

Ah yes, I fixed the ambiguity of `InvariantMonoidal` and added `ContravariantSemigroupal`

satorg reviewed (dismissed):
Looks good to me, thank you!

satorg commented on core/src/main/scala/cats/data/Const.scala:
```diff
@@ -161,10 +171,26 @@ sealed abstract private[data] class ConstInstances extends ConstInstances0 {
         f(fab.getConst, c)
     }
+
+  implicit def catsDataMonoidKForConst[C: Monoid]: MonoidK[Const[C, *]] = new MonoidK[Const[C, *]] {
+    override def empty[A]: Const[C, A] = Const.empty
+    override def combineK[A](x: Const[C, A], y: Const[C, A]): Const[C, A] = x.combine(y)
+  }
+
+  implicit def catsDataSemigroupKForConst[C: Semigroup]: SemigroupK[Const[C, *]] = new SemigroupK[Const[C, *]] {
+    override def combineK[A](x: Const[C, A], y: Const[C, A]): Const[C, A] = x.combine(y)
+  }
```
Just to make sure we are not missing it: wouldn't the following work too?
```suggestion
  implicit def catsDataMonoidKForConst[C: Semigroup]: MonoidK[Const[C, *]] = new MonoidK[Const[C, *]] {
    override def empty[A]: Const[C, A] = Const.empty
    override def combineK[A](x: Const[C, A], y: Const[C, A]): Const[C, A] = x.combine(y)
  }
```
I.e. it looks like even for `MonoidK` it is just `C: Semigroup` that is in use. So since MonoidK <: SemigroupK, wouldn't be it possible to have just one instance of `MonoidK` that demands `C: Semigroup` instead of two separate ones?

joroKr21 replied:
I think you need `Monoid` for `Const.empty` because `Const` always has an element inside.

satorg replied:
Ah yes, you're right. I overlooked it somehow. Thank you for the clarification.

joroKr21 commented on tests/shared/src/test/scala/cats/tests/ConstSuite.scala:
```diff
@@ -120,8 +128,8 @@ class ConstSuite extends CatsSuite {
       const.show.contains(const.getConst.show)
-      assert(const.show === (implicitly[Show[Const[Int, String]]].show(const)))
-      assert(const.show === (const.retag[Boolean].show))
+      assert(const.show === implicitly[Show[Const[Int, String]]].show(const))
```
Probably not

joroKr21 commented on tests/shared/src/test/scala/cats/tests/ConstSuite.scala:
```diff
@@ -120,8 +128,8 @@ class ConstSuite extends CatsSuite {
       const.show.contains(const.getConst.show)
-      assert(const.show === (implicitly[Show[Const[Int, String]]].show(const)))
-      assert(const.show === (const.retag[Boolean].show))
+      assert(const.show === implicitly[Show[Const[Int, String]]].show(const))
```
```suggestion
      assert(const.show === Show[Const[Int, String]].show(const))
```

armanbilge commented on core/src/main/scala/cats/data/Const.scala:
```diff
@@ -174,44 +200,49 @@ sealed abstract private[data] class ConstInstances0 extends ConstInstances1 {
+sealed abstract private[data] class ConstInstances3 extends ConstInstances4 {
 
-  implicit def catsDataFunctorForConst[C]: Functor[Const[C, *]] =
+  @deprecated("Use catsDataTraverseForConst instead", "2.10.0")
```
I think we might want to override `map` in `catsDataTraverseForConst`, for Performance:tm:. Or, so long as we are keeping `ConstFunctor`, we can extend from it to inherit the specialized implementation.

joroKr21 replied:
I made it `final override` because it was also overridden in `ConstApplicative`
